src/auth0/handlers/actionBindings.js

```diff
@@ -0,0 +1,204 @@
+import DefaultHandler from './default';
+import log from '../../logger';
+
+const triggers = [
+  'post-login',
+  'credentials-exchange'
+];
+
+function wait(n) { return new Promise(resolve => setTimeout(resolve, n)); }
+
+async function createBindingWithRetryAndTimeout(client, params, data, retry) {
+  let binding = {};
+  try {
+    if (retry > 0) {
+      binding = await client.createActionBinding(params, data);
+    }
+  } catch (err) {
+    await wait(1000);
+    binding = await createBindingWithRetryAndTimeout(client, params, data, retry - 1);
+  }
+  return binding;
+}
+
+export default class ActionBindingHandler extends DefaultHandler {
+  constructor(options) {
+    super({
+      ...options,
+      type: 'actionBindings'
+    });
+  }
+
+  async getType() {
+    if (this.existing) {
+      return this.existing;
+    }
+
+    // in case client version does not support actions
+    if (
+      !this.client.actionBindings
+      || typeof this.client.actionBindings.getAll !== 'function'
+    ) {
+      return [];
+    }
+    const bindingsResult = [];
+    try {
+      await Promise.all(
+        triggers.map(trigger => this.client.actionBindings
+          .getAll({ trigger_id: trigger, detached: true })
+          .then(b => b.bindings.forEach(binding => bindingsResult.push({ detached: true, ...binding }))))
+      );
+      // get all attached bindings
+      await Promise.all(
+        triggers.map(trigger => this.client.actionBindings
+          .getAll({ trigger_id: trigger, detached: false })
+          .then(b => b.bindings.forEach(binding => bindingsResult.push({ detached: false, ...binding }))))
+      );
+      this.existing = bindingsResult;
+      return this.existing;
+    } catch (err) {
+      if (err.statusCode === 404 || err.statusCode === 501) {
+        return [];
+      }
+      throw err;
+    }
+  }
+
+  async detachBinding(binding, detached) {
+    const params = { trigger_id: binding.trigger_id };
+    const existing = await this.getType();
+    let atacchedList = [];
+    existing.forEach((existingBinding) => {
+      if (!existingBinding.detached) {
+        atacchedList.push({ id: existingBinding.id });
+      }
+    });
+    if (!detached) {
+      atacchedList.push({ id: binding.id });
+    } else {
+      atacchedList = atacchedList.filter(id => id === binding.id);
+    }
+    delete params.binding_id;
+    await this.client.actionBindings.updateList(params, {
+      bindings: atacchedList
+    });
+  }
+
+  // Create Binding creates a atacched binding
+  async createActionBinding(data) {
+    const retries = 10;
+    const params = { trigger_id: data.trigger_id };
+    const actionBinding = { ...data };
+
+    delete actionBinding.created_at;
+    delete actionBinding.detached;
+    delete actionBinding.id;
+    delete actionBinding.trigger_id;
+    delete actionBinding.name;
+    delete actionBinding.version_id;
+    delete actionBinding.action_name;
```

Comment:
Can you explain a little more why we're doing all of these deletes? I'm concerned that we'll be constantly maintaining this list of properties if/when we add more data to the the actions models.

Reply:
I expected these are removed because they are dynamic and should not be sent to the API for create/patching operation.

@luisbritos Maybe we can create a function/whitelist to do the clean-up in one place so it's consistently applied and easier to maintain? Say,
```javascript
const actionBindingExcludeProps = [
    "trigger_id",
    "name"
]

function excludeProps(resource, excludeProps) { ... }

excludeProps(actionBinding, actionBindingExcludeProps);
```

Reply:
thanks Chris! Here I delete these properties from the object because of schema validation for creating an action, for example, create an action will fail with a schema validation if try to pass the action_name as an additional property...
I changed this to create an new object with the properties needed instead of delete the properties of the object.